pkg/proc/fncall.go

```diff
@@ -1016,3 +1086,83 @@ func findCallInjectionStateForThread(t *Target, thread Thread) (*G, *callInjecti
 
 	return nil, nil, notfound()
 }
+
+// debugCallFunction searches for the debug call function in the binary and
+// uses this search to detect the debug call version.
+// Returns the debug call function and its version as an integer (the lowest
+// valid version is 1) or nil and zero.
+func debugCallFunction(bi *BinaryInfo) (*Function, int) {
+	for version := maxDebugCallVersion; version >= 1; version-- {
+		name := debugCallFunctionNamePrefix2 + "V" + strconv.Itoa(version)
+		fn, ok := bi.LookupFunc[name]
+		if ok && fn != nil {
+			return fn, version
+		}
+	}
+	return nil, 0
+}
+
+// debugCallProtocolReg returns the register ID (as defined in pkg/dwarf/regnum)
+// of the register used in the debug call protocol, given the debug call version.
+// Also returns a bool indicating whether the version is supported.
+func debugCallProtocolReg(version int) (uint64, bool) {
+	// TODO(aarzilli): make this generic when call injection is supported on other architectures.
+	var protocolReg uint64
+	switch version {
+	case 1:
+		protocolReg = regnum.AMD64_Rax
+	case 2:
+		protocolReg = regnum.AMD64_R12
+	default:
+		return 0, false
+	}
+	return protocolReg, true
+}
+
+type fakeEntry map[dwarf.Attr]interface{}
+
+func (e fakeEntry) Val(attr dwarf.Attr) interface{} {
+	return e[attr]
+}
+
+func regabiMallocgcWorkaround(bi *BinaryInfo) ([]*godwarf.Tree, error) {
```

Comment:
I took a look just now at the DWARF emitted for runtime.mallocgc on master. It actually doesn't look all that bad, it's getting the incoming params correct at least for the entry point:

```
000000000040a320 <runtime.mallocgc>:
runtime.mallocgc():
/ssd2/go.master/src/runtime/malloc.go:903
  40a320:	cmp    0x10(%r14),%rsp
 ^ "size" now in: { [0: S=0 RAX] }
 ^ "typ" now in: { [0: S=0 RBX] }
 ^ "needzero" now in: { [0: S=0 RCX] }
  40a324:	jbe    40ac39 <runtime.mallocgc+0x919>
  40a32a:	add    $0xffffffffffffff80,%rsp
  40a32e:	mov    %rbp,0x78(%rsp)
  40a333:	lea    0x78(%rsp),%rbp
```

It looks like the return value has an empty location, however (I should add that even without optimization, the return location is not going to be RAX, it will be a stack tmp of some sort). So looks like the workaround is still needed at this point.

Agree that it's hard to always emit correct locations when optimization is turned on.